Sources/Customization/DefaultEventDispatcher.swift

```diff
@@ -88,92 +130,47 @@ open class DefaultEventDispatcher: BackgroundingCallbacks, OPTEventDispatcher {
-            // if we can't batch the events because they are not from the same project or
-            // are being sent to a different url.  we set the batchSizeHolder to batchSize
-            // and batchSize to 1 until we have sent the last batch that couldn't be batched.
-            var batchSizeHolder = 0
-            // the batch send count if the events failed to be batched.
-            var sendCount = 0
-            
-            let failedBatch = { () -> Void in
-                // hold the batch size
-                batchSizeHolder = self.batchSize
-                // set it to 1 until the last batch that couldn't be batched is sent
-                self.batchSize = 1
-            }
             
-            let resetBatch = { () -> Void in
-                if batchSizeHolder != 0 {
-                    self.batchSize = batchSizeHolder
-                    sendCount = 0
-                    batchSizeHolder = 0
+            func removeStoredEvents(num: Int) -> Void {
+                if let removedItem = self.dataStore.removeFirstItems(count: num), removedItem.count > 0 {
+                    // avoid event-log-message preparation overheads with closure-logging
+                    self.logger.d({ "Removed stored \(num) events starting with \(removedItem.first!)" })
+                } else {
+                    self.logger.e("Failed to removed \(num) events")
                 }
-                
             }
+            
             while let eventsToSend: [EventForDispatch] = self.dataStore.getFirstItems(count: self.batchSize) {
-                let actualEventsSize = eventsToSend.count
-                var eventToSend = eventsToSend.batch()
-                if eventToSend != nil {
-                    // we merged the event and ready for batch
-                    // if the bacth size is not equal to the actual event size,
-                    // then setup the batchSizeHolder to be the size of the event.
-                    if actualEventsSize != self.batchSize {
-                        batchSizeHolder = self.batchSize
-                        self.batchSize = actualEventsSize
-                        sendCount = actualEventsSize - 1
-                    }
-                } else {
-                    failedBatch()
-                    // just send the first one and let the rest be sent until sendCount == batchSizeHolder
-                    eventToSend = eventsToSend.first
-                }
+                let (numEvents, batched) = eventsToSend.batch()
                 
-                guard let event = eventToSend else {
-                    self.logger.e(.eventBatchFailed)
-                    resetBatch()
-                    break
+                guard numEvents > 0 else { break }
+                
+                guard let batchEvent = batched else {
+                    // discard events that create invalid batch and continue
+                    removeStoredEvents(num: numEvents)
```

Comment:
i'm not sure this makes sense.  you are saying if batching failed remove all the events that were attempted to batch.

Reply:
In fact, it discards the single event when an invalid event is detected. The comment is misleading. What happens:
- if an invalid one is found while batching, it batches all the valid ones before the invalid one and send it out.
- when trying to batch next, it finds the invalid one at the header. It discards that specific invalid one and continue batching next ones.

I'll fix the comments to avoid confusion.

Reply:
Ok, I got it.  But, numEvents should always be 1, right?  Can you call that out somehow?  You could have passed 1 into here to say remove the offending one.  But, with numEvents, its hidden unless you look at the batch function and figure out where it passes back null.